plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPageSourceProvider.java

```diff
@@ -174,40 +172,52 @@ public ConnectorPageSource createPageSource(
             // Unable to find table schema
             throw new TrinoException(HUDI_FILESYSTEM_ERROR, e);
         }
-        List<HiveColumnHandle> hiveColumns = columns.stream()
+
+        // Convert columns to HiveColumnHandles
+        List<HiveColumnHandle> hiveColumnHandles = columns.stream()
                 .map(HiveColumnHandle.class::cast)
                 .toList();
 
-        List<HiveColumnHandle> regularColumns = hiveColumns.stream()
+        // Get non-synthesized columns (columns that are available in data file)
+        List<HiveColumnHandle> nonSynthesizedColumnHandles = hiveColumnHandles.stream()
                 .filter(columnHandle -> !columnHandle.isPartitionKey() && !columnHandle.isHidden())
                 .collect(Collectors.toList());
-        List<HiveColumnHandle> columnHandles = prependHudiMetaColumns(regularColumns);
 
-        Schema requestedSchema = constructSchema(dataSchema, columnHandles.stream().map(HiveColumnHandle::getName).toList());
+        // The `columns` list could be empty when count(*) is issued, prepending hoodie meta columns allows a non-empty dataPageSource to be returned
+        List<HiveColumnHandle> nonSynthesizedWithHoodieMetaColumnHandles = prependHudiMetaColumns(nonSynthesizedColumnHandles);
 
         TrinoFileSystem fileSystem = fileSystemFactory.create(session);
-        // Only enable predicate pushdown for COW tables
-        boolean enablePredicatePushDown = ((HudiTableHandle) connectorTable).getTableType().equals(HoodieTableType.COPY_ON_WRITE);
+        // Enable predicate pushdown for splits containing only base files
+        boolean isBaseFileOnly = hudiSplit.getLogFiles().isEmpty();
         ConnectorPageSource dataPageSource = createPageSource(
                 session,
-                columnHandles,
+                nonSynthesizedWithHoodieMetaColumnHandles,
                 hudiSplit,
                 fileSystem.newInputFile(Location.of(hudiBaseFileOpt.get().getPath()), hudiBaseFileOpt.get().getFileSize()),
                 dataSourceStats,
                 options.withSmallFileThreshold(getParquetSmallFileThreshold(session))
                         .withVectorizedDecodingEnabled(isParquetVectorizedDecodingEnabled(session)),
-                timeZone, dynamicFilter, enablePredicatePushDown);
+                timeZone, dynamicFilter, isBaseFileOnly);
 
         SynthesizedColumnHandler synthesizedColumnHandler = SynthesizedColumnHandler.create(hudiSplit);
 
+        // Avoid avro serialization if split/filegroup only contains base files
+        if (isBaseFileOnly) {
+            return new HudiBaseFileOnlyPageSource(
+                    dataPageSource,
+                    hiveColumnHandles.isEmpty() ? nonSynthesizedWithHoodieMetaColumnHandles : hiveColumnHandles,
+                    nonSynthesizedWithHoodieMetaColumnHandles,
```

Comment:
nit: `nonSynthesizedWithHoodieMetaColumnHandles` is derived from `hiveColumnHandles`, i.e., they are coupled, not independent, so usually we should only pass in one and let the constructor figure out the derived list (i.e., `allOutputColumns` should derive `physicalDataColumns` in the constructor of `HudiBaseFileOnlyPageSource`, instead of taking both as arguments).  Since this is only temporary, it's fine to be in this shape for now.

Reply:
Agree, this should be cleaned up together with `HoodieFilegroupReader` then. I am only retaining them because i realised that `HoodieFilegroupReader` and `HudiTrinoReaderContext` needed them. 

With so many of these being used all over the place, it's currently quite hard to manage them.